ksqldb-common/src/main/java/io/confluent/ksql/reactive/BasePublisher.java

```diff
@@ -37,6 +37,8 @@
   private Subscriber<? super T> subscriber;
   private long demand;
   private boolean cancelled;
+  private boolean sentComplete;
```

Comment:
I think sentComplete and failure should be volatile as they're accessed outside of a monitor

Reply:
The same is true for subscriber, demand, and cancelled, right? I've marked them all as volatile.

Reply:
Nvm, I misread.

- You're certainly right that `failure` needs to be volatile since it's access from `BasePublisher#subscribe()` before the check that guarantees we're on the right context.
- `subscriber` needs to be volatile since it's accessed from `TestQueryPublisher#hasSubscriber()` and `TestQueryPublisher#sendError()`, which don't have the context guarantee.
- `sentComplete`, `demand`, and `cancelled` don't need to be volatile at the moment since all current usages of these fields are after corresponding context checks, but they all have protected accessors so we're liable to reintroducing similar bugs in the future :-/ Is there a way to guard against this?

I've removed the extra volatiles I added and only left the keyword on `failure` and `subscriber` for now.